vowpalwabbit/parse_args.cc

```diff
@@ -1272,98 +1275,132 @@ VW::LEARNER::base_learner* setup_base(options_i& options, vw& all)
 {
   auto setup_func = all.reduction_stack.top();
   all.reduction_stack.pop();
-  auto base = setup_func(options, all);
+  auto base = std::get<1>(setup_func)(options, all);
 
-  if (base == nullptr)
-    return setup_base(options, all);
+  // returning nullptr means that setup_func (any reduction) was not 'enabled' but
+  // only added their respective command args and did not add itself into the
+  // chain of learners, therefore we call into setup_base again
+  if (base == nullptr) { return setup_base(options, all); }
   else
+  {
+    all.enabled_reductions.push_back(std::get<0>(setup_func));
     return base;
+  }
+}
+
+void register_reductions(vw& all, std::vector<VW::LEARNER::base_learner* (*)(VW::config::options_i&, vw&)>& reductions)
+{
+  std::map<VW::LEARNER::base_learner* (*)(VW::config::options_i&, vw&), std::string> allowlist = {{GD::setup, "gd"},
+      {ftrl_setup, "ftrl"}, {scorer_setup, "scorer"}, {CSOAA::csldf_setup, "csoaa_ldf"},
+      {VW::cb_explore_adf::greedy::setup, "cb_explore_adf_greedy"},
+      {VW::cb_explore_adf::regcb::setup, "cb_explore_adf_regcb"},
+      {VW::shared_feature_merger::shared_feature_merger_setup, "shared_feature_merger"}};
+
+  auto name_extractor = options_name_extractor();
+  vw dummy_all;
+
+  for (auto setup_fn : reductions)
+  {
+    if (allowlist.count(setup_fn)) { all.reduction_stack.push(std::make_tuple(allowlist[setup_fn], setup_fn)); }
+    else
+    {
+      auto base = setup_fn(name_extractor, dummy_all);
+
+      if (base == nullptr)
+        all.reduction_stack.push(std::make_tuple(name_extractor.generated_name, setup_fn));
+      else
+        THROW("fatal: under register_reduction() all setup functions must return nullptr");
+    }
+  }
 }
 
 void parse_reductions(options_i& options, vw& all)
 {
+  std::vector<VW::LEARNER::base_learner* (*)(VW::config::options_i&, vw&)> reductions;
+
   // Base algorithms
-  all.reduction_stack.push(GD::setup);
-  all.reduction_stack.push(kernel_svm_setup);
-  all.reduction_stack.push(ftrl_setup);
-  all.reduction_stack.push(svrg_setup);
-  all.reduction_stack.push(sender_setup);
-  all.reduction_stack.push(gd_mf_setup);
-  all.reduction_stack.push(print_setup);
-  all.reduction_stack.push(noop_setup);
-  all.reduction_stack.push(lda_setup);
-  all.reduction_stack.push(bfgs_setup);
-  all.reduction_stack.push(OjaNewton_setup);
-  // all.reduction_stack.push(VW_CNTK::setup);
+  reductions.push_back(GD::setup);
+  reductions.push_back(kernel_svm_setup);
+  reductions.push_back(ftrl_setup);
+  reductions.push_back(svrg_setup);
+  reductions.push_back(sender_setup);
+  reductions.push_back(gd_mf_setup);
+  reductions.push_back(print_setup);
+  reductions.push_back(noop_setup);
+  reductions.push_back(lda_setup);
+  reductions.push_back(bfgs_setup);
+  reductions.push_back(OjaNewton_setup);
+  // reductions.push_back(VW_CNTK::setup);
```

Comment:
Not related to this PR but do we know why this reduction here is commented out? Should we remove it?

Reply:
that's a good question, no idea

Reply:
I believe there was desire to integrate CNTK into VW at one point before I joined.  I think this can be safely removed.